ft-sdk/src/auth_provider.rs

```diff
@@ -0,0 +1,838 @@
+//! ft_sdk::auth_provider module is only available when the feature "auth-provider" is enabled.
+//! This feature should only be enabled for the auth provider service. Eg email, email-username,
+//! GitHub, Google, etc. Applications that need user data should not enable this feature, and
+//! use the ft_sdk::auth module instead.
+//!
+//! # How Will A Site Create Usernames?
+//!
+//! Usernames are supplied by one of the providers, e.g., email-username provider requires
+//! user to pick a unique username during signup, or GitHub provider provides username. A
+//! site can accept username from only one provider as each provider has different
+//! namespaces for username. If a site wants username feature, the only way to create an account
+//! is via the provider that provides username. If the user wants to log in via another provider,
+//! user will be sent to username provider's "create-username" page. If the user wants to log in
+//! via another provider that provides its own username, the username by that provider will be
+//! used if it is available. If the username is not available, the user will be asked to pick a
+//! new username by going to "create-username" page of the provider that provides username, with
+//! the username as default value.
+//!
+//! # How Will Users Update Their Data?
+//!
+//! ft_sdk::auth creates a bunch of functions that can be used to update user data, name, email,
+//! username etc. The UI will have been provided by the auth provider, or some other generic auth
+//! setting package.
+
+/// In the current session, we have zero or more scopes dropped by different auth
+/// providers that have been used so far. Each auth provider sdk also provides some
+/// APIs that require certain scopes to be present. Before calling those APIs, the
+/// caller can check if the session has enough scopes to call that api. If not, the
+/// caller can request the user to log in again with the required scopes.
+pub struct Scope(pub String);
+
+/// update user data
+pub fn authenticate(
+    conn: &mut ft_sdk::Connection,
+    provider_id: &str,
+    // GitHub may use username as Identity, as user can understand their username, but have never
+    // seen their GitHub user id. If we show that user is logged in twice via GitHub, we have to
+    // show some identity against each, and we will use this identity. Identity is mandatory. It
+    // will be stored as UserData::Identity.
+    //
+    // For the same provider_id, if identity changes, we will only keep the latest identity.
+    identity: &str,
+    data: Vec<ft_sdk::auth::UserData>,
+    user_id: Option<ft_sdk::auth::UserId>,
+    // TODO:
+    // _token: Option<serde_json::Value>,
```

Comment:
Token should be stored in `ft_sdk::auth::UserData::Custom`. Problem is we do not know if there is a single token or more, like a lot of oath provider need auth token and refresh token, and both has to be stored. Custom can handle such stuff.